### DISCUSSION TITLE
Older version of the Centaur

### DISCUSSION THREAD

**Discussion by patchamama** [EdNekebno/DGTCentaurMods#77] [Q&A] [answered]
Very nice and thanks for all the work you have done Ed and others. I have the image that Benny has shared yesterday on my sdcard and I have tried it on the raspberry pi w and 2w but it doesn't work. Among so much code I don't know where to start or what to replace from this image, so I can adapt everything on the old version of centaur, before november 2019 (grey rings)... also I don't have the root or sudo user password from the image but I guess I will have to do everything from scratch with raspbian lite... can you tell me a bit where to fix me to make it work (ie keep original centaur files, old eink screen driver, any extra files?). .... Thanks a lot...

**Comment by wormstein** [comment, marked as answer]
Ed, maybe you can just build a pre alpha release Deb file and publish jete on GitHub considering the situation. Right from master.

Pacha man have you a new berry pi w to image?

**Comment by patchamama** [comment]
Oh, that will be great... I have all, raspberry Pi W and the 2W version... I tryed before to install a develop version using a clone of here... but it never run  with my eink-display (the old version in the centaur)...  I know that there is two versions of eink and drivers... thanks a lot... (Wormstein, I hope that you feel better...).

**Comment by EdNekebno** [comment]
The stuff from Dirk/Benny is a fork that they made an image from, so we don't have root/sudo. I don't mind building a deb file...but I'm pretty sure it won't work in this case. I think the problem is indeed the driver. I have only the later board variant, so that's the one I reverse engineered.

My guess would be then that in November 2019 DGT had some supply problem or an older screen when out of manufacture and they swapped it for the closest possible screen. The driver I use I modified to take care of some DGT weirdness and to enable a hidden feature. Then it might depend patchamama how keen you are on playing around/doing some tests? We can make a likely guess as to the screen type because of the later one. I can point you to the appropriate possible drivers, and the changes you need to make (there is some dgt caused weirdness) to test them. Those drivers come with some example scripts. They will not work properly but if you can get a sign of life from one of them then there's a chance I can produce the modified driver file (I say chance because DGT did not connect the busy line, then everything is also time dependent. But it's a fair chance). That deb file would be much more useful to you :) I'm not sure if that's the kind of thing you want to do...but I figure if you have a pi zero w and a 2w already and are asking then you might be interested in trying :) I can give instructions.

**Reply by patchamama** [reply]
I am very excited about this project. Thanks... I have a bit of experience with python but I ran out of energy after all the unsuccessful tests and it was because of the driver... if you give me some sign or way to start, that is, some driver that I can test until I detect which one the board uses, I would be very grateful... it is a good start... I don't know how to thank you already for what you have achieved... please, give me some idea where to start... (I don't want to have to buy a new centaur or perhaps the controller which I imagine is the eink board itself)...

**Reply by patchamama** [reply]
Master, please, give me some instructions...

**Reply by patchamama** [reply]
My wish is to use the new feautures programed (with a hard work) with my first-old version of the Centaur...

**Reply by EdNekebno** [reply]
What I suggest is install raspbian lite to a card. Then follow these instructions to install the Waveshare drivers. 

https://www.waveshare.com/wiki/2.9inch_e-Paper_Module

The newer centaur uses the epd2in9d . In this package of drivers there are 5 variants of epd2in9 drivers. We know it's not the epd2in9d, but the best guess is it is one of the others.

Once you've cloned the repository and set up everything it says, you want to change the pin definition at the top of this file:

https://github.com/waveshare/e-Paper/blob/master/RaspberryPi_JetsonNano/python/lib/waveshare_epd/epdconfig.py

Change the numbers to the numbers as follows:

    # Pin definition
    RST_PIN         = 12 
    DC_PIN          = 16
    CS_PIN          = 18
    BUSY_PIN        = 13

In the examples folder https://github.com/waveshare/e-Paper/tree/master/RaspberryPi_JetsonNano/python/examples there are some examples for the epd2in9 variants.

epd_2in9_V2_test.py
epd_2in9_test.py
epd_2in9b_V3_test.py
epd_2in9bc_test.py
epd_2in9d_test.py

It's worth trying running each of them. They will definitely not work properly. But if you get a response from just one of them (screen flickers black/white or something similar) and not the others we can guess that is probably it.

If none do anything - it might be because DGT didn't connect the busy pin. Then in the library files: https://github.com/waveshare/e-Paper/tree/master/RaspberryPi_JetsonNano/python/lib/waveshare_epd modify the library for each of the types (similar names) and replace the ReadBusy function. It will have code like:

    def ReadBusy(self):
        while(epdconfig.digital_read(self.busy_pin) == 1):      #  0: idle, 1: busy
            epdconfig.delay_ms(200) 

Just replace whatever the code of the function is with pass, like so:

    def ReadBusy(self):
        pass

Then try the examples again and hope for any sign of life from one of them.

If one works, then we can guess that is the right driver to use, I can then make similar modifications that I made with the other one, and just maybe a deb file with the change will work. 

[As an aside, I think the e-ink screens have built in controllers. I would suspect switching out the screen internally for the later one would probably work. It seems a bit drastic though and it's been ages since I've looked carefully at the internals of the board. I can't remember how it's connected]

**Reply by patchamama** [reply]
Maybe the most simple solution is to change the eink-board-controller.... Ed, can you tell me please, if thisone: https://www.amazon.de/-/en/Waveshare-Resolution-Electronic-Controller-Raspberry/dp/B07J3FHJVP/ref=sr_1_1?dchild=1&keywords=eink%2Bbrett&qid=1635878194&sr=8-1&th=1 run?

**Reply by EdNekebno** [reply]
I think that one is the wrong size. I mentioned it because I think replacing the screen might work - I'm not 100% sure it is a good idea. The problem is that you also need exactly the right minor version of the screen and nowhere seems to list those. It's not something I can advise on :(

**Reply by wormstein** [reply]
Thanks Ed for stepping in. I wish I could send longer replies too

But i think changing the screen might wk because i guess it is attached to Pi not the controller

Other option would be for the other fork to contribute :)) apparently they can work and test an older board.

**Reply by patchamama** [reply]
Thanks Ed... you've given me a lot of detailed information and I think I can manage with all that to do the tests and I cross my fingers for find a driver-controller. Today I will try to get this ready and with this way maybe is possible to get a final and universal version for all DGT Centaurs... thanks a lot... soon I will give some feedback... (Thanks also wormstein and Gute Besserung!)

**Comment by patchamama** [comment]
I have bad news and that is that the e-Paper screen did not respond after following all the instructions given by Ed. In the end I have ordered the 2.9 inch screen from amazon (https://www.amazon.de/gp/product/B072J2LG8N/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&th=1 ) and I will try it in 3 weeks if it works and I will share the results here... Thanks for the well detailed instructions and maybe hopefully something will come up in the next few days....

**Reply by EdNekebno** [reply]
Thanks for letting us know. That's a shame about the drivers :( Exciting about the screen try though.

**Reply by patchamama** [reply]
Thank you Ed!... I finally bought it elsewhere (the 2.9" e-Paper D version) to see if I had it this weekend... I will confirm soon... (https://www.berrybase.de/sensoren-module/displays/epaper-displays/2.9-296-215-128-epaper-display-hat-flexibel-f-252-r-raspberry-pi)

**Reply by EdNekebno** [reply]
Looking at the code for something else I noticed that the epaper in the centaur process is called epaperT5. That makes me wonder if the older screen is a Lilygo T5 and they simply did not rename the class. Then unfortunately I can't find any python drivers but I'm curious when you do this if the old screen has marking on it (like T5_V2.2 on the back shown here: https://nl.aliexpress.com/item/32850386996.html and here: http://www.lilygo.cn/prod_view.aspx?TypeId=50061&Id=1397&FId=t3:50061:3 )

**Reply by OGMalin** [reply]
Visually it looks very much like this: https://www.good-display.com/product/389.html
Maybe this lilygo thing are using the same e-paper.

(Edit: Removed quote)

**Reply by EdNekebno** [reply]
Hmmm...it's possible you're on to something. In the title is a GDEW029T5 that in 2018 was a black white display and now appears to be 4 grayscale. 
Newer version: https://www.e-paper-display.com/products_detail/productId=347.html
Older (i think) spec: https://www.e-paper-display.com/download_detail/downloadsId%3d471.html
Drivers are in C though :(

**Comment by patchamama** [comment]
In the old screen (background tag) only can you read:

E029A01N185Q29
V40B4QL5Q9H100

**Comment by patchamama** [comment]
I wanted to share that I have tested the new epaper screen purchase (> https://www.berrybase.de/sensoren-module/displays/epaper-displays/2.9-296-215-128-epaper-display-hat-flexibel-f-252-r-raspberry-pi) and it did not work on the centaur after changing the screen directly. After some tests, I removed the original display and soldered the controller wires to the gpio pins as instructed by the developer ( https://www.waveshare.com/wiki/2.9inch_e-Paper_Module > User guides of PI > Hardware connection ) and although it now doesn't close the centaur lid properly (due to the size of the controller being separated from the epaper sheet), I was able to get this 2. 9 "d version (I also had to replace the epd2in9d.py and epdconfig.py files with the original files from the developer : https://github.com/waveshare/e-Paper/tree/master/RaspberryPi_JetsonNano/python/lib/waveshare_epd as neither EdNekebno-wormstein's nor dsommerfeld's github files are not worked). The text-image appears with inverted colours but it's ok... Maybe for those who have a first version Centaur, I propose that when the centaur is started, it should go into DGTCentaur mode and then maybe you can connect from the handy via bluetooth (something like pegasus) and you can use the Centaur without a screen, controlling it from the handy with an application like chess for android...

**Reply by EdNekebno** [reply]
Shame :( Not sure if you tried but it's possible just changing the epdconfig might have worked.

dsommerfeld's is almost entirely our code de-structured and made into a build image, so it won't be any different (and I'm finding it increasingly difficult not to get upset about him not following the license terms, it being an open source project and him increasingly making it closed, etc to be honest. But maybe I'm having a grumpy day because yesterday I spent 14 hours straight and only managed to reverse when the charger was plugged in or not! :) ). 

Mobile idea is a good one. Ssshhh...don't tell DGT but one of the things I'd like to do is reverse the pegasus ble protocol. Then your idea would definitely work for that.

**Reply by patchamama** [reply]
At the end if by default the Centaur will enter in DGT mode and the emulation with applications like White Pawn will work well connected, I think with that is enough for us to play the old Centaur in a first step using lichess or another engine... so thinking also maybe we could manage the menues from a web app in the handy connected with the centaur and this way we could see the e-paper screen menu in the handy and try to control everything from it... I will see what I can achieve with this idea in the next week that I will be on vacation... (I'm a little worried about wormstein.... I hope you can't get out of the situation you have quickly and healthy and that it is something temporary... I send you my regards and best wishes....)
